cmd/ssh.go

```diff
@@ -201,195 +202,262 @@ func (cmd *SSHCmd) startProxyTunnel(
 	if err != nil {
-		return fmt.Errorf("failed to connect to %s: %w", serverAddress, err)
+		return fmt.Errorf("failed to connect to outer tunnel: %w", err)
 	}
 	defer conn.Close()
 
-	// Establish an SSH client connection
-	sshConn, channels, requests, err := ssh.NewClientConn(conn, serverAddress, clientConfig)
+	sshConfig := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	clientConn, chans, reqs, err := ssh.NewClientConn(conn, address, sshConfig)
 	if err != nil {
 		return fmt.Errorf("failed to establish SSH connection: %w", err)
 	}
-	client := ssh.NewClient(sshConn, channels, requests)
+	client := ssh.NewClient(clientConn, chans, reqs)
 	defer client.Close()
 
-	// Start an SSH session
-	session, err := client.NewSession()
+	log.Debugf("Outer connection ready")
+	return onReady(client)
+}
+
+// startInnerServer starts the user-level SSH server.
+func startInnerServer(sshClient *ssh.Client, user, workDir string, log log.Logger) error {
+	command := fmt.Sprintf(
+		"/usr/local/bin/devpod helper ssh-server --workdir '%s' --address 0.0.0.0:8023",
+		workDir,
+	)
+	// if user != "root" {
+	// 	command = fmt.Sprintf("su -c \"%s\" '%s'", command, user)
+	// }
+
+	session, err := sshClient.NewSession()
 	if err != nil {
 		return fmt.Errorf("failed to create SSH session: %w", err)
 	}
 	defer session.Close()
 
-	// Configure terminal for interactive shell
-	fd := int(os.Stdin.Fd())
-	if term.IsTerminal(fd) {
-		oldState, err := term.MakeRaw(fd)
-		if err != nil {
-			return fmt.Errorf("failed to set terminal raw mode: %w", err)
-		}
-		defer term.Restore(fd, oldState)
+	var stderr bytes.Buffer
+	session.Stderr = &stderr
 
-		session.Stdout = os.Stdout
-		session.Stderr = os.Stderr
-		session.Stdin = os.Stdin
+	if err := session.Run(command); err != nil {
+		return fmt.Errorf("inner server start failed: %w (stderr: %s)", err, stderr.String())
+	}
+
+	return nil
+}
 
-		termModes := ssh.TerminalModes{
-			ssh.ECHO:          1,
-			ssh.TTY_OP_ISPEED: 14400,
-			ssh.TTY_OP_OSPEED: 14400,
+// waitForInnerServer waits for the inner server to become reachable.
+func waitForInnerServer(ctx context.Context, network tailscale.TSNet, host string, port int) error {
+	address := tailscale.GetURL(host, port)
+	deadline := time.Now().Add(10 * time.Second)
+
+	for time.Now().Before(deadline) {
+		conn, err := network.Dial(ctx, "tcp", address)
+		if err == nil {
+			conn.Close()
+			return nil
 		}
-		err = session.RequestPty("xterm", 80, 24, termModes)
-		if err != nil {
-			return fmt.Errorf("failed to request PTY: %w", err)
+		time.Sleep(500 * time.Millisecond)
+	}
+
+	return fmt.Errorf("inner server not reachable at %s", address)
+}
+
+// waitUntilReachable polls until the given host is reachable via Tailscale.
+func waitUntilReachable(ctx context.Context, ts tailscale.TSNet, host string, log log.Logger) error {
+	const retryInterval = time.Second
```

Comment:
since this is blocking the SSH connection, I think we should poll faster. If we just miss the connection we wait another second, I think this should be something like 100ms. WDYT?